ember-template-assert blog post

nickschot opened the pull request:
TODO:
- [x] Publish addon https://github.com/mainmatter/ember-template-assert
- [x] Reference addon in post
- [x] Conclusion
- [x] Final self-review

emmasofiah2o commented on src/posts/2023-01-13-ember-template-assert.md:
```diff
@@ -0,0 +1,284 @@
+og:
+  image: /assets/images/posts/2022-12-09-sending-emails-from-the-edge-with-rust/og-image.jpg
+tagline: |
+  <p>Sometimes there is the need for development-time only functionality, in this case in the form of an Ember addon. What we'd like to build is an addon that provides us with an <a href="https://api.emberjs.com/ember/4.6/functions/@ember%2Fdebug/assert">assert</a> template helper so that we can use these development time assertions in templates as well.</p>
```
```suggestion
  <p>Sometimes there is the need for development-time-only functionality, in this case in the form of an Ember addon. We'd like to build an addon that provides us with an <a href="https://api.emberjs.com/ember/4.6/functions/@ember%2Fdebug/assert">assert</a> template helper so that we can use these development time assertions in templates as well.</p>
```

emmasofiah2o commented on src/posts/2023-01-13-ember-template-assert.md:
```diff
@@ -0,0 +1,284 @@
+
+The `assert` statement by default runs only for development builds and is
+completely stripped from production. This rather useful functionality does by
+default nog have an equal in an Ember template. This could be a useful addition
```
```suggestion
default not have an equal in an Ember template. This could be a useful addition
```

emmasofiah2o commented on src/posts/2023-01-13-ember-template-assert.md:
```diff
@@ -0,0 +1,284 @@
+The `assert` statement by default runs only for development builds and is
+completely stripped from production. This rather useful functionality does by
+default nog have an equal in an Ember template. This could be a useful addition
+especially for template only components, which have no JavaScript backing class.
```
```suggestion
especially for template-only components, which have no JavaScript backing class.
```

emmasofiah2o commented on src/posts/2023-01-13-ember-template-assert.md:
```diff
@@ -0,0 +1,284 @@
+generate a helper with `ember g helper assert`. In an addon this will generate 3
+files. The file where our implementation will be `/addon/helpers/assert.js`, a
+re-export that will make the helper discoverable by Ember apps
+`/app/helpers/assert.js` and a test file.
```
```suggestion
`/app/helpers/assert.js`, and a test file.
```

emmasofiah2o commented on src/posts/2023-01-13-ember-template-assert.md:
```diff
@@ -0,0 +1,284 @@
+
+Note that from Ember 4.5 onwards it is also possible to use plain functions as
+helpers, removing the need to wrap the function in a `helper()` call. For
+backwards compatibility reasons however, we will keep it for now.
```
```suggestion
backward compatibility reasons, however, we will keep it for now.
```

emmasofiah2o reviewed (commented):
A few comments so far :). Reads well tho, I like that you didn't put too much text in between code.

BobrImperator commented on src/posts/2023-01-13-ember-template-assert.md:
```diff
@@ -0,0 +1,284 @@
+
+Of course, like the AST transform, we only want to remove the helper from the
+build in production environments which we achieve by checking
+`app.isProduction`. We can remove the file from the build by specifying
+`exclude: ['helpers/assert.js]` which matches the file path of the helper itself
+but not the re-export in the app folder.
+
+For that we also need to add the very similar hook called `treeForApp` which
+handles everything in the `app` folder of your v1 ember addon.
```
I'm not sure on the wording, but I feel like adding (sort of) absolute paths to the helper and its re-export would clarify this a bit.

```suggestion
`app.isProduction`. We can remove the file from the build by specifying
`exclude: ['helpers/assert.js]` which matches the file path of the helper itself in the addon directory - 
 `addon/helpers/assert.js`

However we also need to remove the helper from the app tree itself - `app/helpers/assert.js`.
For that we also need to add the very similar hook called `treeForApp` which
handles everything in the `app` folder of your v1 ember addon.
```

nickschot replied:
I've reworded it a bit, but applied the suggestion generally. Thanks!

emmasofiah2o commented:
![og-image ember assert](https://user-images.githubusercontent.com/83696750/213930335-3b260592-fbf2-47fe-8756-7b31a21f7375.jpg)

@nickschot here is your og-image for the post (Idk if the title is final yet?). Hit me up when it's ready for release.

nickschot commented on src/posts/2023-01-13-ember-template-assert.md:
````diff
@@ -0,0 +1,284 @@
+```
+
+This will remove all Mustache statements (any template statement that has
+handlebars syntax like `{{...}}`) with the name `assert`. Since we want to
````
Oh or are you talking about nested helper usage with parens?

emmasofiah2o commented:
Hey @nickschot I can still see some to-dos open at the top: is this ready for release yet?

nickschot commented:
> Hey @nickschot I can still see some to-dos open at the top: is this ready for release yet?

Todo's are done now (though I'll read through it once more). Only thing left is adding the image to the actual post I think?

emmasofiah2o commented:
@nickschot not sure why GitHub decided to merge stuff, but I have checked and it looks like nothing was changed - I think it would be best if you had a look as well 😅

nickschot commented:
> @nickschot not sure why GitHub decided to merge stuff, but I have checked and it looks like nothing was changed - I think it would be best if you had a look as well 😅

Seems there's 2 blogpost files now 🙈 I had renamed the file to a newer date so the order would be correct again. I'll try and fix ASAP.

emmasofiah2o commented:
> > @nickschot not sure why GitHub decided to merge stuff, but I have checked and it looks like nothing was changed - I think it would be best if you had a look as well 😅
> 
> Seems there's 2 blogpost files now 🙈 I had renamed the file to a newer date so the order would be correct again. I'll try and fix ASAP.

Oh I see. Sorry about that! Just ping me when you are ready to release :)

nickschot commented:
> 

Fixed!

nickschot commented:
> > > @nickschot not sure why GitHub decided to merge stuff, but I have checked and it looks like nothing was changed - I think it would be best if you had a look as well 😅
> > 
> > 
> > Seems there's 2 blogpost files now 🙈 I had renamed the file to a newer date so the order would be correct again. I'll try and fix ASAP.
> 
> Oh I see. Sorry about that! Just ping me when you are ready to release :)

Fixed!
